### PR TITLE
 Move overlap_mode into trait level attribute 

### PR DESCRIPTION
**Pull request by spastorino**
r? @nikomatsakis 

Should fix some performance regressions noted on https://github.com/rust-lang/rust/pull/93175

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -487,6 +487,10 @@ declare_features! (
     (incomplete, repr128, "1.16.0", Some(56071), None),
     /// Allows `repr(simd)` and importing the various simd intrinsics.
     (active, repr_simd, "1.4.0", Some(27731), None),
```
Comment (nikomatsakis) [inline compiler/rustc_feature/src/active.rs:489]: This is a "perma-unstable" feature, just for testing purposes...at least thus far...so can you move it up to the group above, e.g. where `rustc_private` and so forth are found? And maybe rename to `rustc_strict_coherence`?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me otherwise

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I think I have a better approach. Go back to the old attribute system, but put the attribute ON THE TRAIT, not the impls. Then check it once, when we start the coherence check, and pass the boolean down to the impls.

**Comment by lqd** [comment]
As we thought yesterday, memoizing the overlap mode should somewhat alleviate these new losses (and turn some of them into small improvements), in addition to regaining the losses from the previous negative impl PR (as the `stm32f4` benchmarks, which had been affected the most, look to be improved now). That'd be easy to try.

(Needless to mention, Niko's idea seems also worthwhile: coherence is hot, and the pairwise overlap checks already have supralinear scaling issues.)

**Review by nikomatsakis — COMMENTED** [review]
This doesn't quite look how I expected. What I expected was that...

* there would be a feature gate for the use of negative impls in coherence
* there would be a `rustc_strict_coherence` attribute that can be placed on traits (gated by the `rustc_attrs` feature gate)

If the feature gate is enabled:

* When we do the overlap check (which appears to be done as part of [building the specialization tree](https://github.com/rust-lang/rust/blob/f8402169aaa12e7bbb9630796a8caec90a3055ca/compiler/rustc_trait_selection/src/traits/specialize/mod.rs#L255-L259)), we could check *once* whether the trait has the attribute:
    * We would pass that information ("is this a strict coherence trait?") into the call to `insert`, which is purely an optimization to avoid recomputing it.
    * `insert` can pass it to [`overlapping_impls`](https://github.com/rust-lang/rust/blob/f8402169aaa12e7bbb9630796a8caec90a3055ca/compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs#L164) 
    * in the overlapping impls check:
    * if the attribute is present, we require that either the types are not unifyable or a negative impl is found
    * if the attribute is not present, we also accept the "legacy" overlap check

If the feature gate is disabled:

* We check for the "strict" attribute but report an error if we see it
* We ignore negative impls in `overlapping_impls`

File: compiler/rustc_middle/src/traits/specialization_graph.rs [hunk 2]
```diff
@@ -31,11 +32,23 @@ pub struct Graph {
...
 }
 
 impl Graph {
     pub fn new() -> Graph {
-        Graph { parent: Default::default(), children: Default::default(), has_errored: false }
+        Graph {
+            parent: Default::default(),
+            children: Default::default(),
+            has_errored: false,
+            overlap_mode: OverlapMode::Stable,
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/traits/specialization_graph.rs:46]: I...hmm. I don't know that I would store this as a field, but I guess it's also ok to do so.

File: compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs [hunk 3]
```diff
@@ -327,8 +331,12 @@ impl GraphExt for Graph {
         loop {
             use self::Inserted::*;
 
-            let insert_result =
-                self.children.entry(parent).or_default().insert(tcx, impl_def_id, simplified)?;
+            let insert_result = self.children.entry(parent).or_default().insert(
+                tcx,
+                impl_def_id,
+                simplified,
+                self.overlap_mode,
```
Comment (nikomatsakis) [inline compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs:338]: ... I was expecting this to be a parameter to the function, I think. Maybe that's annoying to do?

**Review by nikomatsakis — COMMENTED** [review]
r=me but one possible change would be to move `overlap_mode` out so it is not stored as a field; the idea is that it is only used during construction and doesn't matter afterwards, so it doesn't have to be a field

**Comment by nikomatsakis** [comment]
Here is a test we could use ---

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=16337df308de4a32e8117dfe8b9521c3

Without the feature gate, this gets an error.

With the feature gate, I think it should pass, because we have `impl !DerefMut for &T`

**Review by nikomatsakis — APPROVED** [review]
r=me

**Pull request merged**